src/Parsing/TreeSitter.hs

```diff
@@ -48,32 +56,39 @@ parseToPreciseAST
   => Duration
   -> Ptr TS.Language
   -> Blob
-  -> m (Either String (t Loc))
+  -> m (Either SomeException (t Loc))
 parseToPreciseAST parseTimeout language blob = runParse parseTimeout language blob $ \ rootPtr ->
   TS.withCursor (castPtr rootPtr) $ \ cursor ->
     runM (runFail (runReader cursor (runReader (Source.bytes (blobSource blob)) (TS.peekNode >>= TS.unmarshalNode))))
+      >>= either (Control.Exception.throw . UnmarshalFailure) pure
+
+instance Exception TSParseException where
+  displayException = \case
+    ParserTimedOut -> "tree-sitter: parser timed out"
+    IncompatibleVersions -> "tree-sitter: incompatible versions"
+    UnmarshalFailure s -> "tree-sitter: unmarshal failure - " <> show s
 
 runParse
   :: MonadIO m
   => Duration
   -> Ptr TS.Language
   -> Blob
-  -> (Ptr TS.Node -> IO (Either String a))
-  -> m (Either String a)
+  -> (Ptr TS.Node -> IO a)
+  -> m (Either SomeException a)
 runParse parseTimeout language Blob{..} action =
-  liftIO . TS.withParser language $ \ parser -> do
+  liftIO . Control.Exception.try . TS.withParser language $ \ parser -> do
```

Comment:
As with anything catching `SomeException` in `IO`, I think this is recovering from more exceptions than we actually intended it to. E.g. what happens if you’re in `ghci` and hit control C? or the process receives a SIGTERM?

Now that we have this nice `TSParseException` type, can we return that instead?